### DISCUSSION TITLE
Fortls on Windows - Unable to build/debug Fortran files.

### DISCUSSION THREAD

**Discussion by anoban** [krvajal/vscode-fortran-support#864] [Q&A - deprecated see Fortran Discourse] [answered]
**Does fortls support debugging on Windows?**

I have a Windows 11 machine with a complete installation of Visual Studio 2022 with C++ workload (17.5.3), and Visual Studio Code (Version: 1.77.0 (system setup)) with Modern Fortran and Microsoft's C/C++ extensions installed.           

I use OneAPI's ifx as my compiler. The problem is that I cannot find an option to build and debug Fortran programmes within Visual Studio Code. With C programmes, I get a Run C/C++ file and Debug C/C++ file buttons (A "Play" button) in addition to the Run and Debug button in the Activity Bar. But with Fortran programmes, this "Play" button is not available and the Run and Debug button in the Activity Bar doesn't seem to work.              

When I click the blue Run and Debug button, VSCode prompts "Select debugger" and shows Python and "Install an extension for Fortran" as the two available debuggers. (VSCode was invoked from Intel OneAPI command prompt, so all the necessary variables were set). The C/C++ extension works perfectly with C files for compilation and debugging using "tasks.json" and "launch.json". I even tried the example launch.json provided in the Modern Fortran extensions's page and It appears to be configured to use GDB (Which makes me wonder is MSVC even supported?) and I couldn't get it to work either.       

Perhaps this is not the place to say this, but Modern Fortran doesn't recognize "findent.exe" in Windows. Even after adding findent.exe's full path to system PATH and providing it to the extension's "formatter" arg, the extension keeps saying "Installing findent with pip with --User option".

Are there any ways to solve these issues?

**Comment by gnikit** [comment, marked as answer]
The main issue is a duplicate of https://github.com/fortran-lang/vscode-fortran-support/issues/310, I urge you to read the discussions to better understand what's going one.

The secondary issue has been a frequent source of confusion, latest duplicate issue is https://github.com/fortran-lang/vscode-fortran-support/issues/839. You can also find working examples of how to set the setting in the vscode extension's README.

---

Please understand that it is difficult for us to provide one-to-one support, especially for issues that have already answered, since we are all
doing this in our free time. You might want to consider asking these questions to https://fortran-lang.discourse.group/ with the **#help:vscode** or **#help:fortls** tags and instead posting only bug/feature requests on GitHub.

**Comment by anoban** [comment]
Firstly, I appreciate the assistance, thank you. I realize that this project pretty much a one man show (or it appears so) and I did not mean to waste your time. I did look up my issues on google but not here, may be I could have searched in the issues section of this repository. My bad.          

- Got findent to work. Thanks ;)               

I want to add some details here for clarification:            

- By using a hand-coded `tasks.json` and `launch.json` I managed to invoke `gdb-oneapi` from VSCode, but as usual, gdb fails to recognize the symbols (hence breakpoints do not work) in the binary, so the debugging session just runs the binary and exits. It appears like the MSVC dubugger just won't work. 
(This was the `launch.json` that managed to invoke `gdb-oneapi`)     

`{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "(gdb) Launch",
            "type": "cppdbg",
            "request": "launch",
            "program": "${workspaceFolder}/${fileBasenameNoExtension}.exe",
            "args": [],
            "stopAtEntry": true,
            "cwd": "${fileDirname}",
            "environment": [],
            "externalConsole": false,
            "MIMode": "gdb",
            "miDebuggerPath": "C:/Program Files (x86)/Intel/oneAPI/debugger/latest/gdb/intel64/bin/gdb-oneapi.exe",
            "setupCommands": [
                {
                    "description": "Enable pretty-printing for gdb",
                    "text": "-enable-pretty-printing",
                    "ignoreFailures": true
                },
                {
                    "description": "Set Disassembly Flavor to Intel",
                    "text": "-gdb-set disassembly-flavor intel",
                    "ignoreFailures": true
                }
            ]
        }
    ]
}`

- And the `tasks.json` is here:       
         
`{
    "tasks": [
        {
            "type": "cppbuild",
            "label": "Fortran: ifort.exe build active file",
            "command": "ifort.exe",
            "args": [
                "/arch:AVX2",
                "/debug:full",
                "/double-size:128",
                "/free",
                "/iface:cref",
                "/names:lowercase",
                "/noaltparam",
                "/Od",
                "/Qax:CORE-AVX512",
                "/Qdiag-dump",
                "/Qpad",
                "/stand:f18",
                "/warn:all",
                "/Fe${fileDirname}/${fileBasenameNoExtension}.exe",
                "${file}"
            ],
            "options": {
                "cwd": "${fileDirname}"
            },
            "problemMatcher": [
                "$msCompile"
            ],
            "group": {
                "kind": "build",
                "isDefault": true
            },
            "detail": "Task generated by Debugger."
        }
    ],
    "version": "2.0.0"
}`

- Even here, VSCode wasn't able to compile the fortran programme itself (`tasks.json` did not work, saying `Cannot build and debug because the active file is not a C or C++ source file.` may be because of "type": "cppbuild" in the configuration.). I had to compile the programme using the command line, and then "debug" it with `gdb-oneapi` which just executed the binary.       

- A custom-hand coded `launch.json` using MSVC's debugger failed upfront.       

`{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "(Windows) Launch",
            "type": "cppvsdbg",
            "request": "launch",
            "program": "${fileDirname}/${fileBasenameNoExtension}.exe",
            "args": [],
            "stopAtEntry": true,
            "cwd": "${fileDirname}",
            "environment": [],
            "console": "integratedTerminal",
            "preLaunchTask": "Fortran: ifort.exe build active file"
        }
    ]
}`

- In the previous issue (#310) you pointed out, he was able to start and run a debugging session.
But I cannot even find the run/debug buttons. (Images attached for clarification)  
   
Fortran session:          

![Screenshot (15)](https://user-images.githubusercontent.com/99625466/229953808-bfe500ab-d98e-4e80-a05a-bc2759faa4a0.png)

C session:            

![Screenshot (17)](https://user-images.githubusercontent.com/99625466/229955144-d3f97f25-dd61-47b7-a6c2-ed318b469b30.png)

- And debugging inside Visual Studio works perfectly. But the problem is that you have to create new solutions and customize their configurations even for single file programmes. Besides auto-completion, auto-indentation, linting does not work in VS2022.

**I'm in no way criticizing or complaining, and In fact I really appreciate the work you have done here. I just wanted to explicitly point out the issues faced in a Windows machine. I'm learning Fortran as a hobby but fluent in Python and C with can-manage-ish competency in JS. If you have any plans to to fix these or if there are any workaround implementations. (I know that you have pointed out the issue lies mainly in the interaction between Modern Fortran and MS's C/C++ extension)... I'd be delighted to help (provided my knowledge is adequate).**
